mozci/push.py

```diff
@@ -294,14 +296,36 @@ def add(result):
         except MissingDataError:
             pass
 
-        # Gather information from the unittest table. We allow missing data for this table because
-        # ActiveData only holds very recent data in it, but we have fallbacks on Taskcluster
-        # artifacts.
-        try:
-            add(run_query("push_tasks_results_from_unittest", args))
-        except MissingDataError:
-            pass
+        # Let's gather error/results from cache or AD/Taskcluster
+        test_tasks_results = adr.config.cache.get(self.push_uuid, {})
+        if len(test_tasks_results.keys()) > 0:
+            tasks = self.normalized_tasks(tasks)
+            # Let's add cached error summaries to TestTasks
+            for t in tasks:
+                if isinstance(t, TestTask):
+                    error_summary = test_tasks_results.get(t.id)
+                    # Only Test tasks have stored error summary information in the cache
+                    if error_summary:
+                        t._errors = error_summary["errors"]
+                        t._results = error_summary["results"]
+            logger.info(
+                "Fetched tasks errors/results for {} from the cache".format(
+                    self.push_uuid
+                )
+            )
+            return tasks
+        else:
+            # Gather information from the unittest table. We allow missing data for this table because
+            # ActiveData only holds very recent data in it, but we have fallbacks on Taskcluster
+            # artifacts.
+            try:
+                no_cache_args = Namespace(rev=self.rev, branch=self.branch, cache=False)
+                add(run_query("push_tasks_results_from_unittest", no_cache_args))
+            except MissingDataError:
+                pass
+            return self.gather_missing_results(self.normalized_tasks(tasks))
 
+    def normalized_tasks(self, tasks):
```

Comment:
Nit: This is internal, so I'd rename it `_normalize_tasks` (and it could even be static, since it doesn't use any attribute of Push).